# Ultimate Guide to Merging Data in Pandas
## From semi/anti joins to validating data merges

### Introduction
With each data science project or dataset, you want to perform several analyses and create plots to find insights. Often, the raw data never comes in one massive table but in many separate ones. To answer your questions, you should have the skills to join multiple tables into one and then perform operations on them.

You can acquire these skills by learning different kinds of merge operations such as inner join, left and right joins, self and anti joins, merging on indexes, etc.

The goal of this article is that you come away with a strong knowledge of combining data in pandas using precise methods suited for any question you want to ask about your data.

### Pandas merge()
Pandas provide several methods for performing merges on dataframes. Among all the others `merge()` method is the most flexible one. It is a dataframe method and the general syntax is as follows:



`df1.merge(df2, on='common_column')`

When combining tables, there are two terminologies you should be familiar with: The name of the table you use first is called __the left table__ while the other is called __the right table__. In the code snippet above, the left table is `df1` and the right table is `df2`. Also, the verbs join, combine and merge are all used interchangebly.

Now let's see how we perform an inner join:

An inner join will only return rows that have matching values in both tables. During the joining process, you will have to know the common table name which exists in both tables.
<img src='images/inner_join.png'></img>

In [1]:
# Load necessary libraries
import pandas as pd
import numpy as np

In [2]:
# Enable multiple cell outputs
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

In [3]:
# Load necessary dataframes
user_usage = pd.read_csv('data/user_usage.csv')
user_devices = pd.read_csv('data/user_device.csv').drop('user_id', axis='columns')

### Basic Exploration

In [4]:
user_usage.info()
user_devices.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 240 entries, 0 to 239
Data columns (total 4 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   outgoing_mins_per_month  240 non-null    float64
 1   outgoing_sms_per_month   240 non-null    float64
 2   monthly_mb               240 non-null    float64
 3   use_id                   240 non-null    int64  
dtypes: float64(3), int64(1)
memory usage: 7.6 KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 272 entries, 0 to 271
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   use_id            272 non-null    int64  
 1   platform          272 non-null    object 
 2   platform_version  272 non-null    float64
 3   device            272 non-null    object 
 4   use_type_id       272 non-null    int64  
dtypes: float64(1), int64(2), object(2)
memory usage: 10.8+ KB


In [5]:
user_usage.describe()

outgoing_mins_per_month  outgoing_sms_per_month    monthly_mb  \
count               240.000000              240.000000    240.000000   
mean                274.559167               98.968292   3628.602042   
std                 293.745744              111.172685   4486.311513   
min                   0.500000                0.250000      0.000000   
25%                  74.590000               29.030000   1132.230000   
50%                 189.705000               70.775000   1797.975000   
75%                 336.045000              125.627500   4246.617500   
max                1816.630000              906.920000  31146.670000   

             use_id  
count    240.000000  
mean   23285.516667  
std      624.139253  
min    22787.000000  
25%    22888.750000  
50%    22987.500000  
75%    23482.500000  
max    25220.000000

Let's say we have these two tables:

In [6]:
user_usage.head()
user_devices.head()

outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb  use_id
0                    21.97                    4.82     1557.33   22787
1                  1710.08                  136.88     7267.55   22788
2                  1710.08                  136.88     7267.55   22789
3                    94.46                   35.17      519.12   22790
4                    71.59                   79.26     1557.33   22792

use_id platform  platform_version     device  use_type_id
0   22782      ios              10.2  iPhone7,2            2
1   22783  android               6.0    Nexus 5            3
2   22784  android               5.1   SM-G903F            1
3   22785      ios              10.2  iPhone7,2            3
4   22786  android               6.0  ONE E1003            1

This data was downloaded from the KillBiller application. KillBiller was a free service that compared every mobile tariff in UK and Ireland. The first, `user_usage` table contains monthyl statistics of mobile usage of users. `user_devices` table provides details about each users phone such as operating system and phone model.

#### Question 1:
How many users use Android OS and how many use iOS?

To answer this question, we will need the information from both tables. There is one linking attribute between both tables: `use_id`. We will use this column in our merge:

In [7]:
usage_w_os = user_usage.merge(user_devices[['platform', 'use_id']], on='use_id', how='inner')
usage_w_os.sample(5)
print(f'Number of users for each OS: {usage_w_os["platform"].value_counts()}')

outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb  use_id  \
85                    129.85                   66.65     2362.07   22937   
10                    554.41                  150.06     3114.67   22804   
25                    324.27                   91.50      519.12   22830   
150                    57.66                   62.85     1557.33   23036   
52                    128.34                    1.77     7469.79   22881   

    platform  
85   android  
10   android  
25   android  
150  android  
52   android

Number of users for each OS: android    157
ios          2
Name: platform, dtype: int64


It looks like there is a huge difference between the two operating systems in our dataset. 

In the merge above, we used an example of an inner join. In `merge()` function `how` argument is set to `inner` by default so we did not have to write it out. When merging two tables using the `merge()` function, we use `on` argument to specify the common column. If there are multiple, it is also possible to pass a list of columns to the argument and `pandas` will take care of the rest. 

Note that as a right table, I subset the `user_devices` table to exclude irrelevant columns to the question. 

Now, as we explore further, we will notice that number of given users in two datasets is different:

In [8]:
print(f'Dimensions of user_usage table: {user_usage.shape}')
print(f'Dimensions of user_devices table: {user_devices.shape}')
print(f'Dimensions of the joined table: {usage_w_os.shape}')

Dimensions of user_usage table: (240, 4)
Dimensions of user_devices table: (272, 5)
Dimensions of the joined table: (159, 5)


Clearly, matching user IDs in both tables were 159. This means there are user IDs which are in `user_devices` table and not in `user_usage` table and vice versa. So, the next question we want to ask is:

#### Question 2
How many users use Android OS and iOS, including all the users that aren't in `user_usage` table?

We can answer this question by using either a __left__ or a __right__ join. First let's look at the general case of one-sided joins:
<img src='images/left_right_join.png'></img>

A __left join__ will return all of the rows of the left table including matching rows from the right table.
A __right join__ will return all of the rows of the right table including matching rows from the left table. Note that in both types of joins, `pandas` will put a `NaN` value to the cells with no match.

In [9]:
usage_w_os2 = user_usage.merge(user_devices[['platform', 'use_id']], on='use_id', how='right')
usage_w_os2.sample(5)
print(f'Number of users for each OS: {usage_w_os2["platform"].value_counts()}')

outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb  use_id  \
76                    155.71                   10.14     1557.33   22922   
201                      NaN                     NaN         NaN   22869   
36                    300.47                  169.32     3114.67   22855   
110                   200.18                   89.48     5191.12   22975   
166                      NaN                     NaN         NaN   22797   

    platform  
76   android  
201      ios  
36   android  
110  android  
166      ios

Number of users for each OS: android    184
ios         88
Name: platform, dtype: int64


The same result can be achieved using a left join. We can just swap the tables and pass `left` keyword for the `how` parameter:

In [10]:
usage_w_os2 = user_devices[['platform', 'use_id']].merge(user_usage, on='use_id', how='left')
usage_w_os2.sample(5)
print(f'Number of users for each OS: {usage_w_os2["platform"].value_counts()}')

platform  use_id  outgoing_mins_per_month  outgoing_sms_per_month  \
226      ios   23008                      NaN                     NaN   
65       ios   22847                      NaN                     NaN   
104  android   22886                    74.23                    47.4   
222      ios   23004                      NaN                     NaN   
178      ios   22960                      NaN                     NaN   

     monthly_mb  
226         NaN  
65          NaN  
104     1557.33  
222         NaN  
178         NaN

Number of users for each OS: android    184
ios         88
Name: platform, dtype: int64


We use an outer join to return all the rows from both tables regardles they match or not:
<img src='images/outer_join.png'></img>

#### Pandas merge(), suffixes

Let's say we have columns with identical names in both of the tables. Since in our case, we don't have columns with the same name I will just copy out one column into the other:

In [11]:
user_devices['monthly_mb'] = user_usage['monthly_mb']
user_devices.columns
user_usage.columns

Index(['use_id', 'platform', 'platform_version', 'device', 'use_type_id',
       'monthly_mb'],
      dtype='object')

Index(['outgoing_mins_per_month', 'outgoing_sms_per_month', 'monthly_mb',
       'use_id'],
      dtype='object')

Now we have `monthly_mb` column in both tables. Let's see what happends when we join them:

In [12]:
result = user_devices.merge(user_usage, on='use_id')
result.sample(3)

use_id platform  platform_version          device  use_type_id  \
52    22881  android               6.0  HTC Desire 825            1   
14    22813  android               4.4  HTC Desire 510            1   
102   22966  android               6.0        SM-G900F            1   

     monthly_mb_x  outgoing_mins_per_month  outgoing_sms_per_month  \
52        2076.45                   128.34                    1.77   
14        1557.33                   797.06                    7.67   
102       1908.59                   222.32                  224.58   

     monthly_mb_y  
52        7469.79  
14         519.12  
102       3114.67

If we look at the column names, the two columns with the same names have been renamed by adding `_x` and `_y` suffixes. Here, `_x` is for the left table and `_y` is for the right. This is `pandas` default behavior when we don't scpefiy the suffixes. We can control the suffixes easily:

In [13]:
result = user_devices.merge(user_usage, on='use_id', suffixes=('_device', '_stat'))
result.sample(3)

use_id platform  platform_version      device  use_type_id  \
4    22792  android               5.1    SM-G361F            1   
12   22806  android               6.0       A0001            1   
85   22937  android               6.0  HTC One M9            1   

    monthly_mb_device  outgoing_mins_per_month  outgoing_sms_per_month  \
4             3114.67                    71.59                   79.26   
12            1271.39                   283.30                  107.47   
85            3114.67                   129.85                   66.65   

    monthly_mb_stat  
4           1557.33  
12         15573.33  
85          2362.07

It is much better now. Use such suffixes to indicate the columns table of origin when they have identical names. Note that `suffixes` parameter takes a __tuple__ of values, not a list and order does matter!

In [14]:
# Undo all recent changes
user_usage = pd.read_csv('data/user_usage.csv')
user_devices = pd.read_csv('data/user_device.csv').drop('user_id', axis='columns')

#### Pandas merge(), left_on and right_on

There is also a case where you want to join on a common column but the columns have different names. `pandas` provides a handy argument for doing just that without changing the names of the columns. Let's rename the `use_id` columns so that I can show you an example.

In [15]:
user_usage.rename(columns={'use_id': 'user'}, inplace=True)
user_usage.columns
user_devices.columns

Index(['outgoing_mins_per_month', 'outgoing_sms_per_month', 'monthly_mb',
       'user'],
      dtype='object')

Index(['use_id', 'platform', 'platform_version', 'device', 'use_type_id'], dtype='object')

Now, let's combine the tables with an `inner` join:

In [16]:
user_usage.merge(user_devices, left_on='user', right_on='use_id').head()

outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb   user  use_id  \
0                    21.97                    4.82     1557.33  22787   22787   
1                  1710.08                  136.88     7267.55  22788   22788   
2                  1710.08                  136.88     7267.55  22789   22789   
3                    94.46                   35.17      519.12  22790   22790   
4                    71.59                   79.26     1557.33  22792   22792   

  platform  platform_version    device  use_type_id  
0  android               4.3  GT-I9505            1  
1  android               6.0  SM-G930F            1  
2  android               6.0  SM-G930F            1  
3  android               5.1     D2303            1  
4  android               5.1  SM-G361F            1

We use `left_on` and `right_on` parameters of the merge function to specify the names of common columns. You should pass the name of the column from the left table to `left_on` and vice versa. Note that when we use differently named common column, `pandas` will keep both of the columns as duplicates, so don't forget to drop one of them:

In [17]:
user_usage.merge(user_devices, left_on='user', right_on='use_id').drop('user', axis='columns').head()

outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb  use_id  \
0                    21.97                    4.82     1557.33   22787   
1                  1710.08                  136.88     7267.55   22788   
2                  1710.08                  136.88     7267.55   22789   
3                    94.46                   35.17      519.12   22790   
4                    71.59                   79.26     1557.33   22792   

  platform  platform_version    device  use_type_id  
0  android               4.3  GT-I9505            1  
1  android               6.0  SM-G930F            1  
2  android               6.0  SM-G930F            1  
3  android               5.1     D2303            1  
4  android               5.1  SM-G361F            1

In [18]:
# Undo all recent changes
user_usage = pd.read_csv('data/user_usage.csv')
user_devices = pd.read_csv('data/user_device.csv').drop('user_id', axis='columns')

### Pandas merging on indexes

It is also possible to merge on a common index. The syntax for any type of join using indexes is the same. `pandas` is smart enough to know that you are referring to the index:

In [19]:
# Set use_id as indexes
user_usage.set_index('use_id', inplace=True)
user_devices.set_index('use_id', inplace=True)

# Merge on index
user_usage.merge(user_devices, on='use_id', how='outer').head()

outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb platform  \
use_id                                                                         
22787                     21.97                    4.82     1557.33  android   
22788                   1710.08                  136.88     7267.55  android   
22789                   1710.08                  136.88     7267.55  android   
22790                     94.46                   35.17      519.12  android   
22792                     71.59                   79.26     1557.33  android   

        platform_version    device  use_type_id  
use_id                                           
22787                4.3  GT-I9505          1.0  
22788                6.0  SM-G930F          1.0  
22789                6.0  SM-G930F          1.0  
22790                5.1     D2303          1.0  
22792                5.1  SM-G361F          1.0

Things will get a little tricky when we have indexes with different names. Similar to columns, we will have to use `left_on` and `right_on` parameters. Also, there are two more parameters to indicate that we are merging on an index:

In [20]:
# Rename the names of the index
user_usage.index.names = ['index_1']
user_devices.index.names = ['index_2']    # Yes, that's how you do it)))

# Merge on index
user_usage.merge(user_devices, left_on='index_1', left_index=True, right_on='index_2', right_index=True).sample(3)

index_1  outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb  \
22962    22962                   197.43                  115.02    12458.67   
22839    22839                    57.49                   16.73    15573.33   
22939    22939                   221.64                   83.86      519.12   

      platform  platform_version     device  use_type_id  
22962  android               6.0   SM-N910F            1  
22839  android               6.0      A0001            1  
22939  android               4.4  SM-A300FU            1

We set `left_index` and `right_index` to `True` to indicate we are merging on indexes. This also gives duplicate values: one as an index and one as a column. So, don't forget to drop the column when you are merging!

In [21]:
# Undo all recent changes
user_usage = pd.read_csv('data/user_usage.csv')
user_devices = pd.read_csv('data/user_device.csv').drop('user_id', axis='columns')

If the tables you are merging have MultiLevel indexes, you can pass a list of index names to the `on` or `left/right_on` parameters and you will be set!

### Pandas joining a table to itself

There might be cases where you may want to join a table to itself. Joining a table to itself can be useful when you have:
    1. Hierarchical relationships
    2. Sequential relationships
    3. Graph data

General case of joining a table to itself:
<img src='images/self_join.png'></img>

In [ ]:
# Actual code
sequel.merge(sequel, left_on='id', right_on='sequel_id')

As you can see, a self merge can be pretty useful at times. When you use a self join, you often join on different columns of the same table.

### Filtering joins

All the joins we had done so far have returned data from both tables. But sometimes, you would want to filter your table based on values on some other table. `pandas` does not provide this functionality directly. But we can engineer the steps pretty easily.
#### Semi-join Pandas

Semi-joins are useful when you want to subset your data based on observations in other tables. Semi-joins:
    1. Returns the intersection of two tables, similar to an inner join.
    2. Returns only the columns from the __left__ table, not the __right__.
    3. No duplicates
Back to our little tables:
<img src='images/semi_join.png'></img>

As you can see, semi-join is exactly like an inner join but it only includes the results from the left table. Now back to our example data, we want to know the stats for all the users which exist in both tables. But we don't want the final table to contain columns of the `user_devices` table. We can approach this problem in three steps:

In [22]:
# First, perform an inner join to find common use_ids
semi_table = user_usage.merge(user_devices, on='use_id')
semi_table.head(3)

outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb  use_id  \
0                    21.97                    4.82     1557.33   22787   
1                  1710.08                  136.88     7267.55   22788   
2                  1710.08                  136.88     7267.55   22789   

  platform  platform_version    device  use_type_id  
0  android               4.3  GT-I9505            1  
1  android               6.0  SM-G930F            1  
2  android               6.0  SM-G930F            1

After we have the inner join, we want to subset our original left table with the values in the final table:

In [23]:
# Second, create a boolean series
in_both = user_usage['use_id'].isin(semi_table['use_id'])
in_both.sample(5)

84      True
194    False
228    False
238    False
141     True
Name: use_id, dtype: bool

The variable `in_both` now is a `Boolean Series` which we can use to subset the left table

In [24]:
# Finally, subset the left table
users_in_both = user_usage[in_both]
users_in_both.sample(3)
users_in_both.shape

outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb  use_id
57                    74.23                   47.40     1557.33   22886
96                   102.66                   79.84     5191.12   22951
7                     71.59                   79.26      519.12   22795

(159, 4)

The results are as we excpected. If you compare this join to the very first `inner` join we did, it is the exact same table without the columns of the right table:

In [25]:
inner = user_usage.merge(user_devices, on='use_id')
inner.sample(3)
inner.shape

outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb  use_id  \
40                     74.59                  150.59     1038.21   22862   
26                     85.97                   26.94      407.01   22831   
122                   376.59                   79.84     3114.67   22991   

    platform  platform_version     device  use_type_id  
40   android               4.1  HTC One S            1  
26   android               4.1  GT-I8190N            1  
122  android               6.0   SM-G900F            1

(159, 8)

This is a tricky case. Please, reread and practice this section more to really understand it)))

#### Anti-join Pandas

Anti-join is the complete opposite of a semi-join. Anti-join:
    1. Returns the left table, __excluding__ the intersection
    2. Returns only columns from the left table.
<img src='images/anti_join.png'></img>

As an example, let's do the complete opposite of the previous step in semi-join. Let's find out all the users which are only in the `user_usage` table. Step 1:

In [26]:
# Perform a left join
left_joined = user_usage.merge(user_devices, how='left', on='use_id', indicator=True)
left_joined.sample(5)

outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb  use_id  \
130                   124.70                    4.64       11.68   23003   
146                    22.85                   34.54     6577.12   23029   
100                    28.90                   29.42     3114.67   22964   
28                    135.09                   42.02     5191.12   22833   
121                     4.62                   22.94     3114.67   22989   

    platform  platform_version          device  use_type_id _merge  
130  android               5.1  HUAWEI CUN-L01          1.0   both  
146  android               6.0      HTC One_M8          1.0   both  
100  android               6.0        SM-G900F          1.0   both  
28   android               6.0           E6653          1.0   both  
121  android               6.0  HUAWEI VNS-L31          1.0   both

Now, I introduce you to another helpful parameter: `indicator`. When set to `True`, the resulting table of the merge will have one extra column, `_merge`. It indicates whether a row is a result of both tables or one side. As you see, there is `left_only` string in the rows that did not have a match. Now, we want to subset for those rows:

In [27]:
left_only = left_joined.loc[left_joined['_merge'] == 'left_only', 'use_id']
left_only.sample(3)

222    24466
196    23890
214    24399
Name: use_id, dtype: int64

I used `.loc` subsetting to immediaely get the user IDs. Now we can use this pandas `series` to subset the original left table:

In [28]:
left_only_users = user_usage[user_usage['use_id'].isin(left_only)]
left_only_users.sample(5)
left_only_users.shape

outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb  use_id
198                   198.90                   45.22      625.40   23909
234                   558.52                   70.64     3334.28   24963
210                   677.02                  101.86     1132.23   24318
159                   501.92                  165.52     2112.59   23057
183                   190.08                   30.92      369.84   23591

(81, 4)

Again, this is a bit more complicated, so get practicing...

### Pandas concat()

Finally, we covered most of the `merge` function. Now, `pd.concat()` is used for concatenating/adding two or more tables along a particular axis. It also provides some functionalities to manipulate axes. A common way the function is used is when adding tables vertically.

#### Pandas vertical concatenation of dataframes

To add two or more tables vertically, we use `concat()` method, with the argument set to `index`.
<img src='images/vertical_concat.png'></img>

Let's load one the sample tables in three chunks:

In [ ]:
# Load the sample table in three chunks
user_usage_1, user_usage_2, user_usage_3 = [chunk for chunk in pd.read_csv('data/user_usage.csv', chunksize=80)]

# Concatenate the three chunks vertically
user_usage = pd.concat([user_usage_1, user_usage_2, user_usage_3], axis='index')
user_usage.shape

Pay attention to how I used the `chunksize` parameter! It is useful to know as it will be immensely helpful when loading large datasets!

When using `concat()` function, we always pass the tables to be concatenated in a list. To specify the concat axis, we use `axis` parameter. I highly recommend you pass string values to `axis` parameter such as `index` and `columns`. Passing 0 or 1 can be confusing for a lot of people. So, it is a good practice to improve code readability.

`concat()` function does not pay attention to axis labels. If you have numbered indexes in all of your tables, they will be mixed when you concatenate. So, a helpful paramter for resolving the issue is `ignore_index`:

In [ ]:
pd.concat([user_usage_3, user_usage_1, user_usage_2], axis='index', ignore_index=True).head()

This will reset the index and it will count from 0 to n-1.

Another use case for `concat()` is when you have similar tables but with different column names. When you concatenate tables with differing column names, use `join` parameter with a suitable string `value`.
<img src='images/inner_vertical.png'></img>

In [ ]:
# Concat vertically with an inner join
pd.concat([user_usage, user_devices], join='inner', axis='index').head(3)

Using `concat` with `inner` value resulted in a one column dataframe since it is the only common column between tables. Using `left`, `right`, or `outer` would have expected results but values stacked vertically.

### Pandas append()

`.append()` method:
    1. Simplified version of `concat()` method
    2. Supports: `ignore_index`
    3. Does NOT support: `join` and other differences.

Another difference is that it is a DataFrame method. Meaning that it is called on a DataFrame rather than explicitly like `pd.concat`. Simple example:

In [ ]:
user_usage = user_usage_1.append([user_usage_2, user_usage_3], ignore_index=True)
user_usage.shape

### Types of Data Mappings

We saw several types of joins above. In each one, we specifically told the methods the type of join we want and just took a leap of faith that it will return the correct result. However, there are methods to further ensure that we get the correct results. To do the task in code, you have to understand 4 types of data mappings in merges:
    1. One to One
    2. One to Many
    3. Many to One
    4. Many to Many
Let's elaborate. When we use a one_to_one mapping, we expect that our join should return one and only one match for each row in the left table. If we are joing a table with 200 rows with a one_to_one mapping, the result should also contain 200 rows.

Similarly, if it is one_to_many, we allow multiple matches for a single row in our left table. Let's how we implement the idea in code:

In [ ]:
user_usage.merge(user_devices, on='use_id', validate='one_to_one').sample(3)

`merge()` method provides a helpful parameter for validating these kinds of merges. It accepts only four values:
    1. `one_to_one`
    2. `one_to_many`
    3. `many_to_one`
    4. `many_to_many`

What's more, `pandas` will raise a `MergeError` if we validate on an incorrect mapping:

In [ ]:
# Load new sample data
films = pd.read_csv('data/films.csv')
directors = pd.read_csv('data/directors.csv')

In [ ]:
films.merge(directors, on='director_id', validate='one_to_one')

As you can see, the merged raised an error. Let's set it to the correct mapping:

In [ ]:
films.merge(directors, on='director_id', validate='many_to_many').sample(3)

#### Verifying concatenations

It is also possible to validate on axes when using `.concat()` method. It has `verify_integrity` parameter which is set to `False` by default. When set to True, it will raise a ValueError if the axis or columns have duplicate names:

In [ ]:
# Set a new index
directors.set_index('name', inplace=True)

Let's just concatenate the dataframe to itself for the sake of an example:

In [ ]:
pd.concat([directors, directors], verify_integrity=True)

### Conclusion

Finally, the things I wanted to cover in this article are over. We talked about so many different methods for joining data. Believe it or not, there is still more to this topic such as `merge_ordered()` and `merge_asof()` methods but they might be a topic for another article. This just shows how deep and large `pandas` library is and what you can do with it. I intended this article as a one-stop place for combining data. Hope you will use it in practice...